Detecting media change in CD ripper dialog.

lumip opened the pull request:
Closes #7023. I essentially added a thread to `RipCDDialog` that periodically polls `Ripper::MediaChanged()` and issues a reloading of track list / metadata if the return value is `true`.

Added a mutex that controls access to data updates in `RipCDDialog` to avoid potential race conditions from media changes and musicbrainz metadata responses (not sure if those could really happen, but figured I'd rather be on the safe side).

jbroadus commented on src/ripper/ripcddialog.cpp:
```diff
@@ -128,9 +131,15 @@ RipCDDialog::RipCDDialog(QWidget* parent)
-RipCDDialog::~RipCDDialog() {}
+RipCDDialog::~RipCDDialog() {
+  QMutexLocker l(&mutex_);  // ensure now one is holding a lock on the mutex
+                            // when destroying..
```
The thread locks the mutex multiple times per cycle, so it may call this class's methods after this. The Qt child objects aren't destroyed until the base QObject destructor is called. It would be safer to tell the thread to exit and wait for it here.

jbroadus commented on src/ripper/ripcddialog.cpp:
```diff
@@ -324,10 +335,33 @@ QString RipCDDialog::ParseFileFormatString(const QString& file_format,
+    }
+    msleep(500);
+  }
+}
```
Could this be done with a QTimer in RipCDDialog instead of a separate thread? You shouldn't need the mutex in that case.

lumip replied:
Just to confirm whether I understand Qt correctly: There would no longer be need for the mutex because the timer event would be handled by the main event loop, i.e., no concurrency actually happens?

jbroadus replied:
That's correct. You can connect the timer's timeout() signal to a slot and, by default, the slot code will run on the recipient's thread. It looks like there were some new callOnTimeout methods added to QTimer in 5.12, but we need to be compatible with Qt 5.7 until the Debian Stretch EOL.

hatstand commented on src/ripper/ripcddialog.cpp:
```diff
@@ -128,6 +129,9 @@ RipCDDialog::RipCDDialog(QWidget* parent)
     }
   }
+
+  // Start thread for polling
```
This isn't really a thread.

lumip replied:
true, it's not anymore. I'll fix the comment

hatstand commented on src/ripper/ripcddialog.h:
```diff
@@ -83,5 +85,6 @@ class RipCDDialog : public QDialog {
   Ripper* ripper_;
   bool working_;
   CddaSongLoader* loader_;
+  QBasicTimer media_changed_timer_;
```
Why not use `QTimer`? It's more consistent with what we use elsewhere.

lumip replied:
I based this on the implementation of progress polling in `TranscodeDialog`, which uses the `QBasicTimer`. Seems then like I have picked the one example that's inconsistent with the rest of the code. I can of course change this to `QTimer` here, but shouldn't we then change the `TranscodeDialog` as well then?

lumip replied:
I replace the `QBasicTimer` with a `QTimer` now. Should I look into doing the same in `TranscodeDialog` then @hatstand ?

hatstand reviewed (commented):
I don't think the lock is necessary as the timer events happen on the event loop thread as does everything else AFAICT.

jbroadus commented on src/ripper/ripcddialog.cpp:
```diff
@@ -331,3 +335,13 @@ void RipCDDialog::ResetDialog() {
+  if (ripper_->MediaChanged()) {
+    qLog(Debug) << "media changed!";
+    ResetDialog();
+    if (CheckCDIOIsValid()) {
```
It looks like it originally just checked that `cdio_` wasn't null. They should have given it a better name here:
https://github.com/clementine-player/Clementine/commit/4251fd43315c176223764ed568e0a77cffae7571

lumip replied:
That change seems odd to me and there's not rationale for it given in the commit. I would at least now rename it to `EnsureCDIOIsOpened` or, preferrably, move opening of the cdio object back into the class constructor (or at least otherwise treat it internally and not expose this to other classes, ensuring access to resources like this should be encapsulated...)

lumip commented:
Even though I think this should now be done in the sense of having addressed the review comments, I'll convert this to a draft for now while I'm thinking about how to make PR #7030 work - it might be good to combine those two, so it wouldn't make sense to merge this one at this point.

lumip commented:
Made obsolete by #7037
